doc/cli_subprocess_guidelines.md

```diff
@@ -0,0 +1,112 @@
+3. if developers want to locate the target command's executable file, a tool developers can use is `shutil.which` that gives the full executable file path in system, like `shutil.which(git)` returns the full `git.exe` path in windows platform `C:\\Program Files\\Git\\cmd\\git.EXE`. 
+4. if the target cmd is `az`-related, like `az group show --name xxxx`, please use internal function call `cli.invoke(["az", "group", "show", "--name", "xxx"])` from a newly created CLI context to get the target information.
+
+
+### Best practices in subprocess use cases
+
+
+The following sections discuss some secure coding conventions that, when implemented, can help protect CLI applications from command injection vulnerabilities when using `subprocess` module.
+
+#### Proper input validation and sanitization
+
+Input from users or external sources should never be trusted when appending them into `subprocess` cmd. Developers better provide expected patterns to validate and sanitize user input before adding that into subprocess' cmd. 
+Below is an example input sanitizer that only allows alphanumeric characters from the input string.  
+
+```python
+import re
+def sanitize_input(user_input):
+    return re.sub(r'[^a-zA-Z0-9]', '', user_input)
+```
+In a real CLI module scenario, developers better use their corresponding sanitization pattern for reducing the chaining commands risks.
+
+#### Use parameterized cmd statement
+
+A parameterized cmd statement is a way to structure a command starting from defining the command structure first and then provide the parameters that should be inserted into the command separately. 
+In this way, the system will treat the first item as the executable file and the left item as args of that executable application.
+
+```python
+# instead of
+cmd = f"git {user_input}"
+# using 
+cmd = ["git", user_input]
+```
+
+#### Avoid use `shell=True` with user_inputs
+
+When using subprocess module, avoid `shell=True` argument when it comes with cmd containing user inputs. When python executes a command with `shell=True`, the system shell will interpret any string passed in and if this string contains user input without proper sanitization, an attacker can inject malicious commands, as demonstrated in the very beginning of this doc.
+
+
+## Summary
+Ensuring the safety of Azure CLI from command injection under subprocess calling requires an in-depth understanding of these vulnerabilities and also proactive measures to counteract potential exploits. CLI developers can apply the three security practices, if applicable, when using builtin `subprocess`, but it's recommended to use the centralized function `run_cmd` CLI provided, to safeguard CLI modules from command injection attack and for future more accessible security enforcements.
```

Comment:
Should we have a more prescriptive approach and require CLI developers to use the `run_cmd` for any subprocess operation to meet our security standards?

Reply:
We can add a checking rule on those unsupervised subprocess call and give cli core's `run_cmd` as the substitution, with this doc provided as assistance.